iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp

```diff
@@ -722,6 +722,46 @@ void populateVMToCPatterns(MLIRContext *context,
   patterns.insert<ConstOpConversion<IREE::VM::ConstF32Op>>(context);
   patterns.insert<ConstZeroOpConversion<IREE::VM::ConstF32ZeroOp>>(context);
 
+  // ExtF32: Native floating-point arithmetic
+  patterns.insert<CallOpConversion<IREE::VM::AddF32Op>>(context, "vm_add_f32");
+  patterns.insert<CallOpConversion<IREE::VM::SubF32Op>>(context, "vm_sub_f32");
+  patterns.insert<CallOpConversion<IREE::VM::MulF32Op>>(context, "vm_mul_f32");
```

Comment:
200 lines of boilerplate code... this could be generated or assisted with a macro. WDYT?

(fine to land as-is, just wondering how we could make this easier to work on in the future)

Reply:
Definitely worth to think about a better solution. I think it could be possible to send a follow up where we have a single `patterns.intert<>` call. However, we still need to pass all the VM ops that can be converted to EmitC via the pattern. So I don't know if this removes the boilerplate much, which we also have in other places. For example see https://github.com/google/iree/blob/739a596b52226cd0539752ffa7bcce514234afa1/iree/compiler/Dialect/VM/Conversion/StandardToVM/ConvertStandardToVM.cpp#L778-L798

Reply:
We discussed this internally and it should be indeed doable to have a single `patterns.insert` call. The C functions name to call could be either generated via tablegen or one could move the mapping into the CallOpConversion. However, this somehow moves the boilerplate, but does not really reduce it. Your thoughts @ScottTodd?

Reply:
I don't really have much of an opinion on how the code is set up other than
* modifying this set of conversions should be easy (the current structure is easy to edit but risks missing conversions)
* if the set of conversions is 1:1 with no special cases then it could be fully generated, which would make it impossible to miss conversions

We have many other conversion patterns in the project, but these look particularly mechanical.

Reply:
> I don't really have much of an opinion on how the code is set up other than
> 
>     * modifying this set of conversions should be easy (the current structure is easy to edit but risks missing conversions)

That is true, but ...


>     * if the set of conversions is 1:1 with no special cases then it could be fully generated, which would make it impossible to miss conversions

... unfortunately, this is not a full 1:1 set. E.g. constants
https://github.com/google/iree/blob/5de65494a0edea0db932ac5f16a883c1bb8e8cb9/iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp#L652-L654

are not converted via a `CallOpConversion` and also lists `vm.list`as well as other ops need different conversions. For those we do not call functions in `iree/vm/ops.h` and generate more complex code instead. An *(old)* example showing the `vm.list`allocation is vizualized on slide 17 - 20 in the [EmitC slides](https://drive.google.com/file/d/1p2AM7B1beb5oVf_tlTyinnVQoWuodoxl/view?usp=sharing) presented at the ODM last month. So at least for now I think we should land as is.

Reply:
An idea for the straightforward ones is as mentioned use tablegen - we already have a custom tblgen for handling the bytecode encoding, and could extend that to also support the c functions. This makes things declarative in all cases but those requiring extra lowering work. I'm not sure it's the most scalable solution though (especially if we add direct LLVM lowering, or something like an `emitc#` output, etc), so it may not be worth doing.

In general though I'm fine with having lots of one line pattern instantiations like this - it makes it really easy to find/replace them in the code, easy to add new ones (no ambiguity over whether to do it in tablegen or as a pattern or as an instantiation of a templated pattern), and scales linearly with the number of ops (which is fairly bounded).